Add an AEAD encrypting transformer for storing secrets encrypted at rest

smarterclayton opened the pull request:
Tweak the ValueTransformer interface slightly to support additional
context information (to allow authenticated data to be generated by the
store and passed to the transformer). Add a prefix transformer that
looks for known matching prefixes and uses them. Add an AES GCM
transformer that performs AEAD on the values coming in and out of the
store.

Implementation of https://docs.google.com/document/d/1lFhPLlvkCo3XFC2xFDPSn0jAGpqKcCCZaNsBAv8zFdE/edit# and https://github.com/kubernetes/kubernetes/issues/12742

smarterclayton committed:
Add an AEAD encrypting transformer for storing secrets encrypted at rest

Tweak the ValueTransformer interface slightly to support additional
context information (to allow authenticated data to be generated by the
store and passed to the transformer). Add a prefix transformer that
looks for known matching prefixes and uses them. Add an AES GCM
transformer that performs AEAD on the values coming in and out of the
store.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41939)
<!-- Reviewable:end -->

smarterclayton committed:
Update etcd3 storage to leverage storage/value interfaces

Adds context argument which must be set for AES GCM authenticated data
to be passed.

smarterclayton committed:
Allow config to accept a Transformer

smarterclayton committed:
generated: bazel

timothysc commented:
@smarterclayton did you want this for 1.6?

saad-ali commented:
> @smarterclayton did you want this for 1.6?

@timothysc Can you take a look at this PR. It is targeted for 1.7. I'm taking a look too, but I'm a bit out of my depth here.

timothysc commented:
@saad-ali I actually think @kubernetes/sig-auth-pr-reviews should own this one.

saad-ali commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go:
```diff
@@ -45,6 +48,8 @@ type Config struct {
 	Codec  runtime.Codec
 	Copier runtime.ObjectCopier
+	// Transformer allows the value to be transformed prior to persisting into etcd.
+	Transformer value.Transformer
```
How do use the new transformers? Will additional flags be added to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go#L65 in a follow up PR?

smarterclayton replied:
Yes, working with others to do that (proposal is still being discussed for that part).  The transformer interface should be opaque to config and vice versa.

saad-ali commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -55,8 +56,12 @@ func newETCD3Storage(c storagebackend.Config) (storage.Interface, DestroyFunc, e
 	}
+	transformer := c.Transformer
+	if transformer == nil {
+		transformer = value.IdentityTransformer
```
Are there plans to ever make something other than `IdentityTransformer` the out-of-the-box default?

smarterclayton replied:
Maaaybe gzip?

saad-ali reviewed (commented):
Couple of minor questions otherwise LGTM. Hopefully someone from @kubernetes/sig-auth-pr-reviews can review this for final sign off.

timothysc commented:
/cc @jbeda

rjamet commented:
Hi ! I'm a Googler with crypto experience pulled in by @destijl to take a look. Overall, this looks good to me, with a few minor comments:

-  Having the key as AD concerns me. AD just means an attacker cannot make the verifier accept anything that wasn't built with the key, but it doesn't say that the AD isn't leaked in the ciphertext. It might be the case with the primitives you use, but it's not something that an AEAD guarantees in general. If you want to tie ciphertext to an etcd server, you should not use a secret. Either a public etcd server identifier, or a derived version of the etcd key should work. If you don't have good identifiers for a server besides the key, something like HMAC-SHA256(key, 'etcd binding') should also work (since that construction guarantees you can't get the key from the HMAC).

- GCM is fine there, but maybe not ideal. It's easy to forget the constraints later on, but you seem to do things right: you generate random nonces and don't leave it to the caller, and that 2^32 messages 64GB each limit after what the key should be rotated. If possible, programmatic checks for that are usually a good idea. XSalsa20 and Poly1305 would indeed also work here: fundamentally, you just need an AEAD scheme, and there's a lot of them with different constraints. I'm usually partial to AES-CTR + HMAC-SHA256 for its simplicity, but I think it doesn't matter that much overall. 

- "Enable deep integration with sophisticated key store, management, and audit solutions" : Yes please :) We're chatting with @destijl about that.

smarterclayton commented:
> Having the key as AD concerns me. AD just means an attacker cannot make the verifier accept anything that wasn't built with the key, but it doesn't say that the AD isn't leaked in the ciphertext. It might be the case with the primitives you use, but it's not something that an AEAD guarantees in general. If you want to tie ciphertext to an etcd server, you should not use a secret. Either a public etcd server identifier, or a derived version of the etcd key should work. If you don't have good identifiers for a server besides the key, something like HMAC-SHA256(key, 'etcd binding') should also work (since that construction guarantees you can't get the key from the HMAC).

The key in this case is the etcd key location (the path from the root, so `/kubernetes.io/secrets/namespace_name/secret_name`.  Anyone getting access to the cipher text is likely doing so from an etcd dump, and so already has the path at their disposal.

> I'm usually partial to AES-CTR + HMAC-SHA256 for its simplicity, but I think it doesn't matter that much overall.

In this case I was leaning to accepting Go's default AEAD implementation - I agree policy/salsa would be better, but there isn't one in the library and the limitations of the default GCM seemed less risky than using the secret box impl + doing our own AEAD.  I'm fine either way.

rjamet commented:
Oh, I was wrong for the key, sorry about that. For the AEAD, it looks like keeping it as it is would be the simplest option. Thanks for the quick reply !

destijl reviewed (approved):
LGTM, would be good to get this in, we have someone starting on key management and config.

smarterclayton commented:
@k8s-bot kops aws e2e test this

smarterclayton commented:
Note that we can always add new transformer implementations at any time - this is intentionally flexible.

smarterclayton commented:
Applying label as per comment - this doesn't activate this code, so subsequent PRs can easily accommodate requests for change.

php-coder commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes_test.go:
```diff
@@ -0,0 +1,172 @@
+	if err != nil {
+		b.Fatal(err)
+	}
+	// reverse the key order if stale
```
@smarterclayton I don't see in the code how order is changed.

smarterclayton replied:
Yeah, will add in a follow up.  Just a benchmark.

php-coder commented on staging/src/k8s.io/apiserver/pkg/storage/value/transformer.go:
```diff
@@ -0,0 +1,146 @@
+
+// TransformToStorage uses the first transformer and adds its prefix to the data.
+func (t *prefixTransformers) TransformToStorage(data []byte, context Context) ([]byte, error) {
+	transformer := t.transformers[0]
```
Is it possible that the `transformers` is nil here? Should we add a check for that case in `NewPrefixTransformers()`?

smarterclayton replied:
We generally don't guard against passing nil to constructors since it's coding error to get to that point.

php-coder commented on staging/src/k8s.io/apiserver/pkg/storage/value/transformer_test.go:
```diff
@@ -0,0 +1,101 @@
+
+func TestPrefixFrom(t *testing.T) {
+	testErr := fmt.Errorf("test error")
+	transformErr := fmt.Errorf("test error")
```
Having two different errors with the same message is confusing.

smarterclayton replied:
The equality test is different.

php-coder commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -17,13 +17,14 @@ limitations under the License.
+	"k8s.io/apiserver/pkg/storage/value"
 )
 
 func newETCD3Storage(c storagebackend.Config) (storage.Interface, DestroyFunc, error) {
```
I don't see the similar change for the `newETCD2Storage()` that still use hard-coded `etcd.IdentityTransformer`. Was it on purpose?

karataliu replied:
Same question here, any plan to support etcd2?
